### PR TITLE
Feat(EventStoreDB): WIP: Adds EventStoreDB as an experimental backend store

### PR DESCRIPTION
**Pull request by pdt256**
This is an experimental WIP implementation using EventStoreDB as the backend, communicating via gRPC.

~It appears EventStoreDB does not support a global sequence number for events.~
~https://stackoverflow.com/questions/2948523/in-cqrs-event-sourced-do-you-need-a-global-sequence-counter-in-the-event-stor~ Global log file position is now supported in EventStoreDB: https://developers.eventstore.com/clients/dotnet/5.0/reading/#read-all-events

EventStoreDB does not store global sequence number in the same incremental manner as RangeDB. This PR cannot be merged until this difference is resolved.

### CODE REVIEW

**Comment by yordis** [comment]
@pdt256 any updates on this one?

File: provider/eventstore/eventstore.go [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+	client, err := esclient.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create client: %s", err.Error())
+	}
+
+	return client, nil
+}
+
+func (s *eventStore) streamName(name string) string {
+	s.sync.RLock()
```
Comment (yordis) [inline provider/eventstore/eventstore.go:113]: I am not sure why we would lock here, would you mind explaining a bit?

File: provider/eventstore/eventstore.go [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+	defer s.sync.RUnlock()
+	return fmt.Sprintf("%d-%s", s.version, name)
+}
+
+func (s *eventStore) SetVersion(version int64) {
+	s.sync.Lock()
+	s.globalSequenceNumber = 0
+	s.version = version
+	s.sync.Unlock()
```
Comment (yordis) [inline provider/eventstore/eventstore.go:122]: Go question, why can't we do `defer s.sync.Unlock()` here?

Also, why do we need the lock?

File: provider/eventstore/eventstore.go [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+		if err != nil {
+			resultRecords <- rangedb.ResultRecord{
+				Record: nil,
+				Err:    err,
+			}
+			return
+		}
+		defer client.Close()
+
+		readStream, err := client.ReadAllEvents(ctx, direction.Forwards, stream_position.Start{}, ^uint64(0), true)
```
Comment (yordis) [inline provider/eventstore/eventstore.go:145]: This `stream_position.Start{}` makes no sense to me, we should follow up with ES DB folks about it thou

File: provider/eventstore/eventstore.go [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+		if err != nil {
+			if strings.Contains(err.Error(), rpcErrContextCanceled) {
+				resultRecords <- rangedb.ResultRecord{
+					Record: nil,
+					Err:    context.Canceled,
+				}
+				return
+			}
+
+			if strings.Contains(err.Error(), streamNotFound) {
```
Comment (yordis) [inline provider/eventstore/eventstore.go:155]: This is bad, we should talk to ES DB folks ... checking error message name stuff 😨

File: provider/eventstore/eventstore.go [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+				Record: nil,
+				Err:    err,
+			}
+			return
+		}
+		defer client.Close()
+
+		readStream, err := client.ReadAllEvents(ctx, direction.Forwards, stream_position.Start{}, ^uint64(0), true)
+		if err != nil {
+			if strings.Contains(err.Error(), rpcErrContextCanceled) {
```
Comment (yordis) [inline provider/eventstore/eventstore.go:240]: Same here, we should be able to use `errors.Is/2`

File: provider/eventstore/eventstore.go [hunk 6]
```diff
@@ -0,0 +1,636 @@
...
+				resultRecords <- rangedb.ResultRecord{
+					Record: nil,
+					Err:    context.Canceled,
+				}
+				return
+			}
+
+			if strings.Contains(err.Error(), streamNotFound) {
```
Comment (yordis) [inline provider/eventstore/eventstore.go:248]: And here

**Review by yordis — CHANGES_REQUESTED** [review]
Overall looking amazing, but scares me the way we are checking errors

File: provider/eventstore/eventstore.go (hunk 1 above)
Comment (pdt256) [reply provider/eventstore/eventstore.go]: locks have been removed

File: provider/eventstore/eventstore.go (hunk 3 above)
Comment (pdt256) [reply provider/eventstore/eventstore.go]: outdated, and resolved

**Comment by pdt256** [comment]
Blocked by: https://github.com/EventStore/EventStore/issues/3226

**Pull request merged**